src/net/proto.lua

```diff
@@ -0,0 +1,186 @@
+local sendDataByPeer = {}
+
+local function allocSendData()
+  -- the schema of data sent to a peer
+  local data = {
+    cmd=PROTOCMD_SEND,
+    coordsById={},
+  }
+  return data
+end
+
+local function getSendData(peer)
+  -- XXX JCD peer can also be an enet.host for building up broadcasts
+  local data = sendDataByPeer[peer]
+  if not data then
+    data = allocSendData()
+    sendDataByPeer[peer] = data
+  end
+  return data
+end
+
+local function clearSendData(peer)
+  sendDataByPeer[peer] = nil
+end
+
+local function msgFromData(data)
+  return serpent.dump(data)
+end
+
+local function dataFromMsg(msg)
+  local ok, data = serpent.load(msg)
+  if not ok then
+    logger:error("failed to decode message: %s", msg)
+    return nil
+  end
+  -- TODO validate schema of data
+  return data
+end
+
+function proto:connect(client, host, port)
+  local serverAddr = string.format("%s:%u", host, port)
+  return client:connect(serverAddr)
+end
+
+function proto:join(server, id)
+  local data = {cmd=PROTOCMD_JOIN, version=VERSION, id=id}
+  local msg = msgFromData(data)
+  server:send(msg)
+end
+
+function proto:leave(server, id)
+  local data = {cmd=PROTOCMD_LEAVE, id=id}
+  local msg = msgFromData(data)
+  server:send(msg)
+end
+
+function proto:clientDisconnected(peer)
+  joinedPeers[peer] = nil
+  for id, client in pairs(clientsById) do
+    if client == peer then
+      love.event.push('netmanPlayerLeft', id)
+      clientsById[id] = nil
+    end
+  end
+end
+
+function proto:serverDisconnected(server)
+  joinedPeers[server] = nil
+end
+
+function proto:welcome(id)
+  local client = clientsById[id]
+  if not client then
+    logger:error("could not find connection for player %s", id)
+    return
+  end
+  joinedPeers[client] = true
+  love.event.push('netmanPlayerJoined', id)
+  local data = {cmd=PROTOCMD_WELCOME, id=id, text="welcome to the server!"}
+  local msg = msgFromData(data)
+  client:send(msg)
+end
+
+function proto:reject(peer, id, reason)
+  local data = {cmd=PROTOCMD_REJECT, id=id, text=reason}
+  local msg = msgFromData(data)
+  peer:send(msg)
+end
+
+function proto:prepare(peer, data)
+  local sendData = getSendData(peer)
+  if data.type == netman.SEND_COORD then
+    -- we only care about the most recent coord; overwrite whatever is here
+    sendData.coordsById[data.id] = data.coord
+  else
+    logger:fatal("unrecognized send request type: %s", data.type)
+  end
+end
+
+function proto:send(peer)
+  if not joinedPeers[peer] then
+    -- we're not joined yet; don't send any data
+    return
+  end
+  local data = getSendData(peer)
+  local msg = msgFromData(data)
+  peer:send(msg)
+  clearSendData(peer)
+end
+
+function proto:broadcast(server)
+  local data = getSendData(server)
+  local msg = msgFromData(data)
+  for peer in pairs(joinedPeers) do
+    peer:send(msg)
+  end
+  clearSendData(server)
+end
+
+function proto:recv(peer, msg)
+  local data = dataFromMsg(msg)
+  if not data then
+    return
+  end
+
```

Comment:
```suggestion
if data.version != VERSION then
  logger:error("client version %u, but server running %u",
                    data.version, VERSION)
  peer:disconnect(netman.DISCONNECT_BAD_VERSION)
```

Reply:
data.version is usually nil, so we'd be killing clients pretty fast. xD

Reply:
I just wanted to be cool ⏩ but I 🍂